### PR TITLE
[libc++] Protect the libc++ implementation from CUDA SDK's `__noinline__` macro

### PR DESCRIPTION
**Pull request by gribozavr**
The CUDA SDK contains an unfortunate definition for the `__noinline__`
macro. This patch works around it by using `__attribute__((noinline))`
instead of `__attribute__((__noinline__))` on CUDA. We are still waiting
for a long-term resolution to this issue in NVIDIA/cccl#1235.

### CODE REVIEW

**Comment by jyknight** [comment]
Has anyone filed a bug against CUDA to stop doing this?

File: libcxx/test/libcxx/system_reserved_names.gen.py
Comment (ldionne) [inline libcxx/test/libcxx/system_reserved_names.gen.py:162]:
```suggestion
#define __noinline__ SYSTEM_RESERVED_NAME
```

**Comment by EricWF** [comment]
I'm OK taking this patch, but I would like us to file a bug against CUDA first.

**Review by ldionne — COMMENTED** [review]
On the one hand, working around this appears to be fairly simple and we have precedent for this for other macros. On the other hand, NVIDIA apparently acknowledges the risk for conflicts when they define this macro according to the SO answer:

```
/* gcc allows users to define attributes with underscores,
   e.g., __attribute__((__noinline__)).
   Consider a non-CUDA source file (e.g. .cpp) that has the
   above attribute specification, and includes this header file. In that case,
   defining __noinline__ as below  would cause a gcc compilation error.
   Hence, only define __noinline__ when the code is being processed
   by a  CUDA compiler component.
*/
```

Fixing this here doesn't provide much of an incentive for them to fix their SDK, but being incompatible with a major C++ standard library implementation would. Do we have anyone from Nvidia around here who could enlighten us on the status of that bug report?

File: libcxx/test/libcxx/system_reserved_names.gen.py
Comment (EricWF) [inline libcxx/test/libcxx/system_reserved_names.gen.py:162]: I believe we want to make this macro be defined as SYSTEM_RESERVED_NAME to prove we're not actually using it.

File: libcxx/include/__undef_macros
Comment (EricWF) [inline libcxx/include/__undef_macros:30]: Please add a comment about where the problematic macro comes from (cuda).

**Comment by philnik777** [comment]
Given that you say yourself that this causes problems with lots of other libraries, I think it would be better to fix the CUDA SDK - especially since it's a really simple fix for them.

**Review by EricWF — APPROVED** [review]
~LGTM once the bots come back.~

There has been a bunch of discussion about if we want this. I probably LGTM'ed this to soon.

Maybe we should get some CUDA builders?

File: libcxx/include/__undef_macros
Comment (gribozavr) [reply libcxx/include/__undef_macros:30]: We do that in `system_reserved_names.gen.py`, this file does not have comments for other macros. Do you really want me to break the pattern? Or should I also add comments for other macros in this file?

**Comment by mordante** [comment]
I assume this change means we need to push and pop our macros where we use `__noinline__` via our macro. Or am I mistaken? This patch doesn't seem to do that. Before approving I would like to know what the status of fixing this in the CUDA SDK is.

**Comment by gribozavr** [comment]
> This patch doesn't seem to do that.

That infrastructure was already in place, it was introduced by https://reviews.llvm.org/D147356. The only missing piece was the push/pop bracketing in `__config` itself, which this patch does.

> I think it would be better to fix the CUDA SDK - especially since it's a really simple fix for them.

Yes, we are filing a bug for the CUDA SDK. However the fix will arrive in future SDK versions, the ones currently in use are already conflicting. So ideally libc++ will keep this workaround until those existing SDK versions become obsolete.

**Comment by Artem-B** [comment]
Clang already works around a handful of cases where we ran into this: https://github.com/llvm/llvm-project/commits/main/clang/lib/Headers/cuda_wrappers

@jrhemstad What would be a proper channel to ask NVIDIA's nvcc/CUDA owners to help with that? I can file a bug directly with NVIDIA, but this is a public issue and it would be great to track it publicly, too.

File: libcxx/test/libcxx/system_reserved_names.gen.py
Comment (gribozavr) [reply libcxx/test/libcxx/system_reserved_names.gen.py]: Indeed, fixed.

**Comment by jyknight** [comment]
Should the fix be done by adding a `clang/lib/Headers/cuda_wrappers/__config` wrapper instead, so that all the CUDA `__noinclude__` hacks are kept in one place?

**Comment by Artem-B** [comment]
> Should the fix be done by adding a `clang/lib/Headers/cuda_wrappers/__config` wrapper instead, so that all the CUDA `__noinline__` hacks are kept in one place?

If that's the only place where the issue pops up, then it's an option. So far that's what we did for libstdc++ workarounds.

File: libcxx/include/__undef_macros
Comment (EricWF) [reply libcxx/include/__undef_macros:30]: I don't have a strong preference.

**Comment by Artem-B** [comment]
> Otherwise, in case CUDA says something like "yeah we don't really care", then we'd be bending backwards to support something bad that another vendor is doing in complete awareness and I would push back on doing that on the basis that it's not a good precedent to set nor a sane long term direction for libc++.

See. https://github.com/NVIDIA/thrust/issues/1703 

The way I read it, the fact that NVCC hacks around this issue is a strong indication that it's not going to be fixed any time soon. And we still need to deal with existing CUDA versions anyways, so what NVIDIA is going to do about it is almost a moot point.

File: libcxx/include/__undef_macros
Comment (gribozavr) [reply libcxx/include/__undef_macros:30]: OK, then I'll keep the existing pattern.

**Review by philnik777 — CHANGES_REQUESTED** [review]
It actually looks like nvc++ isn't working around the issue. They simply parse GNU attributes slightly differently: https://godbolt.org/z/MKKf4Wdan. I think we can fix this all by simply using `[[__gnu__::__noinline__]]` instead. This works on every compiler we support except Clang 16 in C++03 mode, so I don't think that's a huge problem. If anybody complains that the CodeGen is different if cuda is enabled, we can tell them to complain to the cuda sdk people and hopefully it'll be fixed at some point.

**Comment by gribozavr** [comment]
@jyknight I agree with @Artem-B 's analysis. To work around this problem, we need to hide CUDA's `__noinline__` macro definition everywhere where `_LIBCPP_NOINLINE` is used (not just where it is defined) - and it is in potentially in every libc++ header. The only solution that achieves that involves adding `__noinline__` to `_LIBCPP_PUSH_MACROS`, `_LIBCPP_POP_MACROS`, and to `__undef_macros` - like this PR.

**Comment by philnik777** [comment]
@gribozavr I'm pretty sure there are tests for this kind of thing. Clang has to both expand macros in C++11 attribute and ignore unknown ones. There may not be a test for this specific case, but for the general case I expect tests. I think we can avoid the preprocessor problem by having the conditional as `#if !defined(_LIBCPP_CXX03_LANG) || (defined(_LIBCPP_CLANG_VER) && _LICPP_CLANG_VER >= 1700)`. It's not pretty, but we avoid having another macro we have to push/pop and we can remove the macro entirely after the next release branch anyways.

**Comment by gribozavr** [comment]
@philnik777 

> I'm pretty sure there are tests for this kind of thing.

I can't find anything using the pattern `::__attribute`. Could you confirm that it is indeed intentionally supported?

> I think we can avoid the preprocessor problem by having the conditional as `#if !defined(_LIBCPP_CXX03_LANG) || _LICPP_CLANG_VER >= 1700`.

Yes, that indeed works.

> we can remove the macro entirely after the next release branch anyways.

Why? Because the next release will not support compilers that don't support the `[[__gnu__::__noinline__]]` syntax?

**Comment by Artem-B** [comment]
Looks like `__noinline__` as a macro, as it's defined right now is going to be problematic.

Perhaps the best choice is to permanently undef CUDA's `__noinline__` or redefine it as __cuda_noinline__ and let users deal with that if they want/need to. Otherwise we'll be playing this game of whack-an-__inline__-conflict forever.

`__noinline__` appears to be relatively rarely used. I see a single reference in [pytorch](https://github.com/pytorch/pytorch/blob/32b928e582f90978c6e647eb1f27831506dc1d9b/aten/src/ATen/native/cuda/IGammaKernel.cu#L373)  and there are a handful in NVIDIA's [core libraries' tests](https://github.com/search?q=repo%3ANVIDIA%2Fcccl%20__noinline__&type=code) . I think getting rid of them should be doable.

At the moment, only few libc++ headers use _LIBCPP_NOINLINE, so we can introduce a temporary build knob which would temporarily  undef `__inline__` in the wrappers for `__config` and `string`, but which would need to be explicitly enabled by the user.

So, transition would work roughly like this:
* temporary undef `__noinline__` in wrappers for `__config` and `string`
* Introduce a knob to completely undef `__noinline__` and disable the workarounds. Keep it enabled initially.
* Switch the knob default to off for the next major clang release.
* Remove the knob in the next major release.

If it's too much hassle, we can keep adding the wrappers. It's a hassle, but it would probably work OK, considering that it's not used all that often.

WDYT, all?

**Comment by gribozavr** [comment]
> There can be basically any string after the `::` that clang has to accept

But then it wouldn't have the desired noinline effect, right?

**Comment by philnik777** [comment]
@gribozavr Do you plan to work on this?

File: libcxx/include/__config
Comment (EricWF) [inline libcxx/include/__config:49]: Is there an open bug we can reference here? Just to mark the noinline code with? (hopefully to encourage us to remove it).

**Review by EricWF — COMMENTED** [review]
So libc++ only references `__noinline__` twice. 

Given that `__noinline__` expands to `__attribute__((noinline))` couldn't we just:

```c++
#ifdef __noinline__
#define _LIBCPP_NOINLINE __noinline__
#else
#define _LIBCPP_NOINLINE __attribute__((__noinline__))
#endif
```

**Review by EricWF — APPROVED** [review]
> ```c++
> 
> #define _LIBCPP_NOINLINE _
> ```
OK, I was wrong about this, because we don't know if CUDA has been included yet.

The original solution is very much appropriate for the stated goals, and goes with existing practice. Once we get the formatted pass I say we land it.

It's clear to me that of the proposed solution, the one that keeps `_LIBCPP_NOINLINE` having semantics  seems the best.
I don't understand the objections that suggest we break the macro, even though we have the infra not to.


Let's do the useful thing with this patch and then continue our beef with nvidia elsewhere. We agree they're naughty, so lets add them to the naughty list.

**Comment by EricWF** [comment]
@gribozavr Could you please update this so the bots are happy, then we can land this?

**Review by philnik777 — CHANGES_REQUESTED** [review]
@EricWF Please don't just overrule people. I don't see why we should make this comparatively complex change if we can just slightly change how we write the attribute. I don't see any reason fur us to make our implementation work super well with cude if there doesn't seem to be much incentive from nvidia to make things work well for other libraries. I'm also not convinced that this parallels with other macros we push and pop, since this isn't about conformance. This is only about a not necessarily super large optimization.

**Comment by EricWF** [comment]
@philnik777 Nobody is overruling you. I appreciate you taking the time to spell out your concerns.

> I don't see any reason fur us to make our implementation work super well with cude if there doesn't seem to be much incentive from nvidia to make things work well for other libraries.

I would say the incentive for us is to help our users, who have asked us for this. 

> I'm also not convinced that this parallels with other macros we push and pop, since this isn't about conformance.

Could you say more? For me the parallel is "Expansion of this macro breaks libc++".

Fundamentally this change simply adds a name to a list. I don't share your concern regarding comparative complexity.

My concern is shipping an implementation where `_LIBCPP_NOINLINE` is a lie when it doesn't need to be. I believe having source code behave as written is a better outcome than having the compiler silently swallow the diagnostic. 

You've requested changes on this CL. Could you please reiterate the actionable items you would like taken?

**Review by mordante — CHANGES_REQUESTED** [review]
I'm also against merging.
@ldionne asked for an issue filed at the CUDA project. This has not been done.

I feel it sets a bad precedent when we make changes when other members in the ecosystem don't want to play nicely. Especially, since they were aware it clashed with existing GCC practice. Instead of using `__cuda_noinline__` they decided to introduce a macro that breaks existing code.

The fix is quite simple, but it has a maintenance burden for us. Every time we use our noinline macro we have to remember to push and pop our macros. Which most likely is only flagged by the CI so needs another CI run. Our CI resources are limited and they slow down development, at least for me it does regularly.

**Comment by EricWF** [comment]
Is deeply saddened by the communities unwillingness to help our users.

**Comment by EricWF** [comment]
> Our CI resources are limited and they slow down development, at least for me it does regularly.

Thanks for the feedback on the CI. Could you please file a bug about it where a meaningful conversation can take place. We're working on improving the CI experience and feedback from developers is super valuable.

**Comment by gribozavr** [comment]
> @ldionne asked for an issue filed at the CUDA project. This has not been done.

Above in the thread @Artem-B has linked to a bug about this that was filed against the Nvidia Thrust library, and was diagnosed as "This issue is unique to clang": NVIDIA/thrust#1703

**Comment by philnik777** [comment]
> @philnik777 Nobody is overruling you. I re-requested your review. I appreciate you taking the time to spell out your concerns.

Saying this can be merged while others having concerns/alternative ways forward that aren't addressed does feel like overruling to me.

> > I don't see any reason fur us to make our implementation work super well with cude if there doesn't seem to be much incentive from nvidia to make things work well for other libraries.
> 
> I would say the incentive for us is to help our users, who have asked us for this.

Users have asked that our code works with cuda, which is the case with my proposed alternative, so I don't see how this doesn't address that.

> > I'm also not convinced that this parallels with other macros we push and pop, since this isn't about conformance.
> 
> Could you say more? For me the parallel is "Expansion of this macro breaks libc++".
> 
> Fundamentally this change simply adds a name to a list. I don't share your concern regarding comparative complexity.
> 
> My concern is shipping an implementation where `_LIBCPP_NOINLINE` is a lie when it doesn't need to be. I believe having source code behave as written is a better outcome than having the compiler silently swallow the diagnostic.

I don't think it doesn't behave as written. `noinline` is basically a hint that a function shouldn't be inlined. Not adding it doesn't break anything.

> You've requested changes on this CL. Could you please reiterate the actionable items you would like taken?

I would like to explore simply using a different spelling to avoid the hard error.




> > @ldionne asked for an issue filed at the CUDA project. This has not been done.
> 
> Above in the thread @Artem-B has linked to a bug about this that was filed against the Nvidia Thrust library, and was diagnosed by an Nvidia employee as "This issue is unique to clang" and closed: [NVIDIA/thrust#1703](https://github.com/NVIDIA/thrust/issues/1703)

This bug has been filed against thrust, but the actual problem is the cuda sdk IIUC. I'm not aware of any bug filed against that.

**Comment by gribozavr** [comment]
> > Above in the thread @Artem-B has linked to a bug about this that was filed against the Nvidia Thrust library, and was diagnosed by an Nvidia employee as "This issue is unique to clang" and closed: [NVIDIA/thrust#1703](https://github.com/NVIDIA/thrust/issues/1703)
> 
> This bug has been filed against thrust, but the actual problem is the cuda sdk IIUC. I'm not aware of any bug filed against that.

I would be happy to file a bug in a bug tracker of your choice - could you let me know where you prefer?

Note that that specific response comes with a person with GitHub bio "Lead for CUDA C++ Core Libraries (CCCL)", so there is a high chance that a bug against the CUDA SDK will be triaged by them too.

**Comment by ldionne** [comment]
If the CUDA SDK truly dismisses this interoperation issue with Clang and GCC, they probably don't care that their customers are able to use those compilers / standard libraries alongside the CUDA SDK, which would be a reason to push back on doing anything here.

I agree we should strive to make our users lives better, but blindly working around this is harmful in the long term for everyone, since the badly-behaved vendor will just keep doing this sort of stuff and we'll only accumulate more and more workarounds. We need to give a good reason for that vendor to change things, and then we can make our users lives better in the meantime while they are transitioning away from the macro.

I will file a bug report against CCCL myself.

**Comment by EricWF** [comment]
The previous release of libc++ did not use `__noinline__`. 

If we ship this release without a fix, we will break many of our users.

**Comment by ldionne** [comment]
This is the bug report: https://github.com/NVIDIA/cccl/issues/1235.

I suggest we hold off on additional discussion here until we have news from them, which should hopefully be quick.

**Comment by mordante** [comment]
> > Our CI resources are limited and they slow down development, at least for me it does regularly.
> 
> Thanks for the feedback on the CI. Could you please file a bug about it where a meaningful conversation can take place. We're working on improving the CI experience and feedback from developers is super valuable.

The main issue is the CI takes a long time and when a lot of people have patches the feedback cycle can take quite some time limiting to one iteration per day. This is mainly an issue when the patch affects system specific behaviour. For example, locales. I don't think there is a good solution for that. Typically I prefer to work on these things in the weekend, then the CI is less busy.

The main point I want to make is that these work-arounds have a cost for us. It was not the intention to bash our CI.

**Comment by miscco** [comment]
Hey, unfortunately the cccl team has no sway over the cuda SDK headers

The feedback that I got internally is that this has been discussed in the past and there is no plan to address this. From the point of the CUDA SDK it is the implementation so using those identifiers is allowed.

**Comment by Artem-B** [comment]
So, to sum it up:
* clang needs to be able to compile CUDA sources w/ libc++ with N previous CUDA SDK versions that will not ever be fixed. This makes the question whether future CUDA SDK versions will be fixed moot.
* https://github.com/llvm/llvm-project/pull/74123 partially works around the issue for now. At least it allows to avoid the outright compialtion failures in *all* CUDA compilations with libc++. It's far from perfect, but it buys us some time to figure a better solution. It does depend on the implementation detail and is potentially fragile, but `__noinline__` is used rather infrequently, so I don't think it will be a big issue in practice.

I propose landing  the workaround  in #74123 if this PR needs more time.

**Comment by jrhemstad** [comment]
> The feedback that I got internally is that this has been discussed in the past and there is no plan to address this.

To be clear, this is not the definitive answer. 

As I mentioned in https://github.com/NVIDIA/cccl/issues/1235#issuecomment-1871402785, I am working on communicating with the relevant internal teams to come to a better solution that everyone can be happy with.

**Comment by dkolsen-pgi** [comment]
The easiest workaround that I can think of is to use `noinline` instead of `__noinline__` when compiling CUDA code.  To do that, just change this code in `__config`:
```
#  if __has_attribute(__noinline__)
#    define _LIBCPP_NOINLINE __attribute__((__noinline__))
#  else
#    define _LIBCPP_NOINLINE
#  endif
```
to:
```
#  if defined(__CUDACC__) || defined(__CUDA_ARCH__) || defined(__CUDA_LIBDEVICE__)
#    define _LIBCPP_NOINLINE_ATTR_NAME noinline
#  else
#    define _LIBCPP_NOINLINE_ATTR_NAME __noinline__
#  endif
#  if __has_attribute(_LIBCPP_NOINLINE_ATTR_NAME)
#    define _LIBCPP_NOINLINE __attribute__((_LIBCPP_NOINLINE_ATTR_NAME))
#  else
#    define _LIBCPP_NOINLINE
#  endif
```
(The condition in the first `#if` matches the condition in CUDA's `crt/host_defines.h` where `__noinline__` is defined as a macro.)

No pushing/popping of macros or creating wrapper headers.  No maintenance burden going forward if `_LIBCPP_NOINLINE` is used in more places.

This is just a workaround, not a fix.  Compilation can still fail if CUDA programs also define `noinline` as a macro.  But I expect that to be extremely rare, because defining the macro `noinline` would likely break the CUDA headers.

I have no objection to withholding any workaround until NVIDIA shows some interesting in improving the situation on the CUDA side of things.  But if there is progress in that area and libc++ wants to check in a workaround on their side, I think this is the way to go.

(Full disclosure:  I work for NVIDIA, though not in the CUDA or NVCC organizations, so I don't have any direct say in what CUDA does.  I am working with @jrhemstad on "communicating with the relevant internal teams to come to a better solution that everyone can be happy with.")

**Comment by mordante** [comment]
> The easiest workaround that I can think of is to use `noinline` instead of `__noinline__` when compiling CUDA code. To do that, just change this code in `__config`:
> 
> ```
> #  if __has_attribute(__noinline__)
> #    define _LIBCPP_NOINLINE __attribute__((__noinline__))
> #  else
> #    define _LIBCPP_NOINLINE
> #  endif
> ```
> 
> to:
> 
> ```
> #  if defined(__CUDACC__) || defined(__CUDA_ARCH__) || defined(__CUDA_LIBDEVICE__)
> #    define _LIBCPP_NOINLINE_ATTR_NAME noinline
> #  else
> #    define _LIBCPP_NOINLINE_ATTR_NAME __noinline__
> #  endif
> #  if __has_attribute(_LIBCPP_NOINLINE_ATTR_NAME)
> #    define _LIBCPP_NOINLINE __attribute__((_LIBCPP_NOINLINE_ATTR_NAME))
> #  else
> #    define _LIBCPP_NOINLINE
> #  endif
> ```
> 
> (The condition in the first `#if` matches the condition in CUDA's `crt/host_defines.h` where `__noinline__` is defined as a macro.)
> 
> No pushing/popping of macros or creating wrapper headers. No maintenance burden going forward if `_LIBCPP_NOINLINE` is used in more places.
> 
> This is just a workaround, not a fix. Compilation can still fail if CUDA programs also define `noinline` as a macro. But I expect that to be extremely rare, because defining the macro `noinline` would likely break the CUDA headers.
> 
> I have no objection to withholding any workaround until NVIDIA shows some interesting in improving the situation on the CUDA side of things. But if there is progress in that area and libc++ wants to check in a workaround on their side, I think this is the way to go.
> 
> (Full disclosure: I work for NVIDIA, though not in the CUDA or NVCC organizations, so I don't have any direct say in what CUDA does. I am working with @jrhemstad on "communicating with the relevant internal teams to come to a better solution that everyone can be happy with.")

Thanks for working on trying to solve this inside NVIDIA!
I like your work-around better than the current proposal, thanks for the suggestions.

**Comment by Artem-B** [comment]
> use noinline instead of __noinline__ when compiling CUDA code.

SGTM.

We should probably add a comment that this specifically targets the `__noinline__` macro definition in CUDA's `/crt/host_defines.h`. If CUDA headers' implementation changes in the future this workaround may break and would have to grow version-dependent variants.

We may make the workaround a bit simpler and more robust by directly checking if `__noinline__` is defined. However that opens a possibility of us applying the workaround if/when `__noinline__` may be defined as a macro for reasons other than CUDA headers.

**Comment by philnik777** [comment]
The last solution actually sounds really nice to me. It's basically "if you do UB by defining `__noinline__` we're claiming `noinline` for ourselves".

**Comment by ldionne** [comment]
I'd be OK with the suggested workaround since folks are actively escalating this with the Cuda SDK. So the whole patch could be as simple as:

```c++
#  if defined(__CUDACC__) || defined(__CUDA_ARCH__) || defined(__CUDA_LIBDEVICE__)
#    define _LIBCPP_NOINLINE __attribute__((noinline)) // LINK TO BUG REPORT
#  elif __has_attribute(__noinline__)
#    define _LIBCPP_NOINLINE __attribute__((__noinline__))
#  else
#    define _LIBCPP_NOINLINE
#  endif
```

@gribozavr Are you able to update this PR accordingly? Ideally we'd land this before we cut the LLVM 18 branch next Tuesday.

File: libcxx/include/__config
Comment (mordante) [inline libcxx/include/__config:1214]: I really would like a comment this is a temporary work-around. 
Just  to make clear we still expect the CUDA SDK team to solve this properly.
Without this comment it seems that this is a permanent solution.

Comment (gribozavr) [reply libcxx/include/__config:1217]: Done.

**Review by ldionne — APPROVED** [review]
LGTM. Thanks for the patch, and thanks everyone for the discussion.

I'll merge this later today. In the meantime, I will ping @philnik777, @mordante and @EricWF to make sure they have a chance to see this quickly and express any concerns.

**Review by philnik777 — APPROVED** [review]

**Review by mordante — APPROVED** [review]

**Comment by Artem-B** [comment]
> > The feedback that I got internally is that this has been discussed in the past and there is no plan to address this.
> 
> To be clear, this is not the definitive answer.
> 
> As I mentioned in [NVIDIA/cccl#1235 (comment)](https://github.com/NVIDIA/cccl/issues/1235#issuecomment-1871402785), I am working on communicating with the relevant internal teams to come to a better solution that everyone can be happy with.

The bug I've opened with NVIDIA has just been closed  as a "not NVIDIA's problem". :-(

**Comment by jrhemstad** [comment]
> The bug I've opened with NVIDIA has just been closed  as a "not NVIDIA's problem". :-(

Yeah, this isn't right. I have commitment from the compiler team to fix this. This sounds like the bug was closed by someone who doesn't know the context.

**Comment by Artem-B** [comment]
> Which bug report? Do you have a link?

It is, unfortunately, siloed. AFAICT there's no publlic way to report CUDA bugs other than loudly complaining on NVIDIA's forums.

**Pull request merged**
